src/t_set.c

```diff
@@ -477,7 +477,7 @@ void setTypeConvert(robj *setobj, int enc) {
         sds element;
 
         /* Presize the dict to avoid rehashing */
-        dictExpand(d, setTypeSize(setobj));
+        dictExpand(d, setTypeSize(setobj) + 1);
```

Comment:
that's true for the calls made from `setTypeAddAux`, but in the calls made from rdb.c we actually know how many items we're gonna push next.
maybe we should pass an optional hint to this function by the caller?

Reply:
Good point. In `rdb.c` the dict is sometimes presized by an extra call to dictTryExpand after setTypeConvert. These will benefit from giving a presize hint to convert.

To log an OOM error in rdbLoadObject, we can let setTypeConvert return ERR for OOM errors and use it when we change this rdb code:

```C
                    setTypeConvert(o,OBJ_ENCODING_HT);
                    if (dictTryExpand(o->ptr,len) != DICT_OK) {
                        rdbReportCorruptRDB("OOM in dictTryExpand %llu", (unsigned long long)len);
```

Currently, a failed dictExpand is ignored inside setTypeConvert. In scenarios like repeated SADD, this will lead to expand triggered later by dictAddRaw, which is a smaller expand, so we can end up with a smaller hash table, which is a fine way to avoid OOM I guess. Thus, a failed dictExpand is not a failed setTypeConvert, so the return value is only a hint back about the success of pre-sizeing...

Reply:
> Currently, a failed dictExpand is ignored inside setTypeConvert

won't it OOM panic? (i.e. the difference between dictTryExpand and dictExpand is that it won't panic on OOM).
we need to keep panicking if it happens at runtime, and handle it gracefully only in RESTORE command.

Reply:
OK (you're right, I realized and edited..)